torchbenchmark/models/nvidia_deeprecommender/nvtrain.py

```diff
@@ -0,0 +1,403 @@
+                                      nl_type=self.args.non_linearity_type,
+                                      is_constrained=self.args.constrained,
+                                      dp_drop_prob=self.args.drop_prob,
+                                      last_layer_activations=not self.args.skip_last_layer_nl)
+    os.makedirs(self.args.logdir, exist_ok=True)
+    self.model_checkpoint = self.args.logdir + "/model"
+    self.path_to_model = Path(self.model_checkpoint)
+    if self.path_to_model.is_file():
+      print("Loading model from: {}".format(self.model_checkpoint))
+      self.rencoder.load_state_dict(torch.load(self.model_checkpoint))
+  
+    if not self.args.silent:
+      print('######################################################')
+      print('######################################################')
+      print('############# AutoEncoder Model: #####################')
+      print(self.rencoder)
+      print('######################################################')
+      print('######################################################')
+  
+    if self.args.use_cuda:
+      gpu_ids = [int(g) for g in self.args.gpu_ids.split(',')]
+      if not self.args.silent:
+        print('Using GPUs: {}'.format(gpu_ids))
+      
+      if len(gpu_ids)>1:
+        self.rencoder = nn.DataParallel(self.rencoder,
+                                   device_ids=gpu_ids)
+      self.rencoder = self.rencoder.cuda()
+  
+    if self.args.optimizer == "adam":
+      self.optimizer = optim.Adam(self.rencoder.parameters(),
+                                  lr=self.args.lr,
+                                  weight_decay=self.args.weight_decay)
+    elif self.args.optimizer == "adagrad":
+      self.optimizer = optim.Adagrad(self.rencoder.parameters(),
+                                lr=self.args.lr,
+                                weight_decay=self.args.weight_decay)
+    elif self.args.optimizer == "momentum":
+      self.optimizer = optim.SGD(self.rencoder.parameters(),
+                            lr=self.args.lr, momentum=0.9,
+                            weight_decay=self.args.weight_decay)
+      self.scheduler = MultiStepLR(self.optimizer, milestones=[24, 36, 48, 66, 72], gamma=0.5)
+    elif args.optimizer == "rmsprop":
+      self.optimizer = optim.RMSprop(self.rencoder.parameters(),
+                                lr=self.args.lr, momentum=0.9,
+                                weight_decay=self.args.weight_decay)
+    else:
+      raise  ValueError('Unknown optimizer kind')
+  
+    self.t_loss = 0.0
+    self.t_loss_denom = 0.0
+    self.denom = 0.0
+    self.total_epoch_loss = 0.0
+    self.global_step = 0
+  
+    if self.args.noise_prob > 0.0:
+      self.dp = nn.Dropout(p=self.args.noise_prob)
+
+  def DoTrain(self):
+  
+    self.rencoder.train()
+    #if self.args.optimizer == "momentum":
+    #  self.scheduler.step()
+  
+    for i, mb in enumerate(self.data_layer.iterate_one_epoch()):
```

Comment:
the training function will be more complex of course.. but the same trick.  you need to end up with the model, so you can just construct an object of the train benchmark for that purpose.  but you'll have to duplicate the DoTrain code and cut out some parts.

- the for loop should not exist, instead you'll just want to intercept (in pdb) the first output from it, save that to a pickle, then prepare it in your __init__() so that it's already unpickled and on cuda and in dense format, then can just be used directly by the train function

Reply:
I understand the pickle part. Could you please articulate what should be in the `train()` function (suppose the data is already unpickled)?